### PR TITLE
Refactoring of RAM

### PR DESCRIPTION
**Pull request by b-scholz**
This represents the first step of refactoring the intermediate representation of RAM. Several subsequent steps of refactoring are required. RAM mimics a similar design like the AST of Souffle.

Following further steps for a cleaner design are necessary:
1) RamRelation needs to be split into a reference and a relation declaration 
2) RamTranslationUnit is to be generated by the AST Translator
3) Introduce RAM transformers for levelling 
4) Introduce RAM analyses for indexing

### CODE REVIEW

File: src/RamCondition.h [hunk 1]
```diff
@@ -31,138 +31,208 @@
...
+        const RamAnd& other = static_cast<const RamAnd&>(node);
+        return getLHS() == other.getLHS() && getRHS() == other.getRHS();
+    }
 };
 
-/** abstract class for binary operation/relations on values */
+/**
+ * Binary constraint
+ */
+// TODO: rename to RamConstraint
```
Comment (mmcgr) [inline src/RamCondition.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamCondition.h [hunk 2]
```diff
@@ -31,138 +31,208 @@
...
+/** Not existence check for a relation */
 class RamNotExists : public RamCondition {
-    /** the relation to be accessed */
+protected:
+    /* Relation */
     RamRelation relation;
 
-    /** the restricted fields -- null if undefined */
+    /** Pattern -- nullptr if undefined */
+    // TODO: rename to argument
```
Comment (mmcgr) [inline src/RamCondition.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamRelation.h [hunk 3]
```diff
@@ -32,29 +32,46 @@ namespace souffle {
...
-    bool computed;
-    bool output;
-    bool btree;
-    bool brie;
-    bool eqrel;
 
-    bool isdata;
-    bool istemp;
+    /** Relation qualifiers */
+    // TODO: Simplify interface
```
Comment (mmcgr) [inline src/RamRelation.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamRelation.h [hunk 4]
```diff
@@ -32,29 +32,46 @@ namespace souffle {
...
+    /** Arity, i.e., number of attributes */
     unsigned arity;
+
+    /** Name of attributes */
     std::vector<std::string> attributeNames;
+
+    /** Type of attributes */
     std::vector<std::string> attributeTypeQualifiers;
+
+    /** TODO: legacy, i.e., duplicated information */
```
Comment (mmcgr) [inline src/RamRelation.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamRelation.h [hunk 5]
```diff
@@ -32,29 +32,46 @@ namespace souffle {
 
 /**
  * A RAM Relation in the RAM intermediate representation.
+ * TODO: Make RamRelation a sub-class of RAM node.
```
Comment (mmcgr) [inline src/RamRelation.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamCondition.h [hunk 6]
```diff
@@ -212,34 +276,94 @@ class RamNotExists : public RamCondition {
...
+        const RamNotExists& other = static_cast<const RamNotExists&>(node);
+        return getRelation() == other.getRelation() && equal_targets(values, other.values);
+    }
 };
 
-/** check whether a given relation is empty or not*/
+/**
+ * Emptiness check for a relation
+ */
+// TODO: Rename to RamEmptyCheck
```
Comment (mmcgr) [inline src/RamCondition.h]: TODOs should be in the form` // TODO (#issueid): text` or `// TODO (userid): text`

File: src/RamTranslationUnit.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "DebugReport.h"
+#include "ErrorReport.h"
+#include "RamAnalysis.h"
+#include "RamProgram.h"
+#include "SymbolTable.h"
+
+#include <map>
+#include <memory>
+
```
Comment (mmcgr) [inline src/RamTranslationUnit.h]: `#include <string>`

File: src/RamOperation.h [hunk 8]
```diff
@@ -25,32 +25,38 @@
...
+ * Abstract class for a relational algebra operation
+ */
 class RamOperation : public RamNode {
 protected:
     /** the nesting level of this operation */
     size_t level;
 
-    /** conditions that is checked for each obtained tuple */
+    /**
+     * conditions that is checked for each obtained tuple
```
Comment (mmcgr) [inline src/RamOperation.h]: Condition

File: src/RamOperation.h [hunk 9]
```diff
@@ -63,67 +69,109 @@ class RamOperation : public RamNode {
...
+            condition = map(std::move(condition));
+        }
+    }
+
+protected:
+    /** Check equality */
+    bool equal(const RamNode& node) const override {
+        assert(dynamic_cast<const RamOperation*>(&node));
+        const RamOperation& other = static_cast<const RamOperation&>(node);
+        if (getCondition() != nullptr && other.getCondition() != nullptr) {
```
Comment (mmcgr) [inline src/RamOperation.h]: I'm probably missing something, but aren't all these ifs equivalent to a single `return getCondition() == other.getCondition();`?

Is the first if body meant to be comparing pointers or objects pointed to?

File: src/RamAnalysis.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/RamAnalysis.h]: No - they are replicas of the Ast intermediate representation. The copyright is still hold by Oracle and not us.

File: src/ParserDriver.cpp [hunk 11]
```diff
@@ -31,9 +31,10 @@ ParserDriver::ParserDriver() : trace_scanning(false), trace_parsing(false) {}
 
 ParserDriver::~ParserDriver() = default;
 
-std::unique_ptr<AstTranslationUnit> ParserDriver::parse(const std::string& f, FILE* in, bool nowarn) {
+std::unique_ptr<AstTranslationUnit> ParserDriver::parse(
+        const std::string& f, FILE* in, SymbolTable& s, ErrorReport& e, DebugReport& d) {
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ParserDriver.cpp]: done

File: src/ParserDriver.cpp [hunk 12]
```diff
@@ -51,9 +52,10 @@ std::unique_ptr<AstTranslationUnit> ParserDriver::parse(const std::string& f, FI
     return std::move(translationUnit);
 }
 
-std::unique_ptr<AstTranslationUnit> ParserDriver::parse(const std::string& code, bool nowarn) {
+std::unique_ptr<AstTranslationUnit> ParserDriver::parse(
+        const std::string& code, SymbolTable& s, ErrorReport& e, DebugReport& d) {
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ParserDriver.cpp]: done

File: src/ParserDriver.cpp [hunk 13]
```diff
@@ -72,14 +74,15 @@ std::unique_ptr<AstTranslationUnit> ParserDriver::parse(const std::string& code,
 }
 
 std::unique_ptr<AstTranslationUnit> ParserDriver::parseTranslationUnit(
-        const std::string& f, FILE* in, bool nowarn) {
+        const std::string& f, FILE* in, SymbolTable& sym, ErrorReport& e, DebugReport& d) {
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ParserDriver.cpp]: done

File: src/ParserDriver.h [hunk 14]
```diff
@@ -30,6 +30,7 @@ class AstRelation;
 class AstType;
 class AstProgram;
 class SymbolTable;
+class DebugReport;
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ParserDriver.h]: done

File: src/ParserDriver.h [hunk 15]
```diff
@@ -60,12 +61,14 @@ class ParserDriver {
 
     bool trace_scanning;
 
-    std::unique_ptr<AstTranslationUnit> parse(const std::string& f, FILE* in, bool nowarn = false);
-    std::unique_ptr<AstTranslationUnit> parse(const std::string& code, bool nowarn = false);
+    std::unique_ptr<AstTranslationUnit> parse(
+            const std::string& f, FILE* in, SymbolTable& s, ErrorReport& e, DebugReport& d);
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/ParserDriver.h]: done

File: src/RamCondition.h [hunk 16]
```diff
@@ -31,138 +31,208 @@
...
+
+    /** Create clone */
+    RamCondition* clone() const override = 0;
 };
 
-/** class for matching a string with a pattern */
+/**
+ * Conjunction
+ */
+// TODO: rename to RAMConjunction
```
[the remark this answers is not in the document]
Comment (b-scholz) [reply src/RamCondition.h]: Done.

File: src/RamCondition.h (hunk 1 above)
Comment (b-scholz) [reply src/RamCondition.h]: Done.

File: src/RamCondition.h (hunk 2 above)
Comment (b-scholz) [reply src/RamCondition.h]: Done.

File: src/RamCondition.h (hunk 6 above)
Comment (b-scholz) [reply src/RamCondition.h]: Done.

File: src/RamOperation.h (hunk 9 above)
Comment (b-scholz) [reply src/RamOperation.h]: Indeed this was a semantic error. It should have been 
```
return *getCondition() == *other.getCondition();
```

File: src/RamTranslationUnit.h (hunk 7 above)
Comment (b-scholz) [reply src/RamTranslationUnit.h]: Done.

File: src/RamRelation.h (hunk 3 above)
Comment (b-scholz) [reply src/RamRelation.h]: Done.

File: src/RamRelation.h (hunk 4 above)
Comment (b-scholz) [reply src/RamRelation.h]: Done.

File: src/RamRelation.h (hunk 5 above)
Comment (b-scholz) [reply src/RamRelation.h]: Done.

**Pull request closed without merging**